src/Ocelot/WebSockets/Middleware/WebSocketsProxyMiddleware.cs

```diff
@@ -117,6 +117,16 @@ private static async Task Proxy(HttpContext context, string serverEndpoint)
                 }
             }
 
+            // Only Uris starting with 'ws://' or 'wss://' are supported in System.Net.WebSockets.ClientWebSocket
+            if (!serverEndpoint.StartsWith("ws"))
+            {
+                var scheme = context.Items.DownstreamRequest().Scheme;
+                Logger.LogWarning($"Invalid scheme has detected which will be replaced! Scheme '{scheme}' of the downstream '{serverEndpoint}'.");
+                serverEndpoint = serverEndpoint
```

Comment:
question: is it really the role of this method to transform the scheme? 
I think that if the configuration isn't right, an exception should be raised.
There's nothing more confusing for a dev than to see in a config file a scheme defined as "http" and then find that it's executed as "wss", a magic trick!

Reply:
> @ArtRoman commented [on Sep 15](#issuecomment-1721577694)

Maybe because of that ☝️ WS protocol is based on http, and developer confuses too often.
@ArtRoman Explain please once again.

Reply:
@RaynaldM If execution goes to WebSocketsProxyMiddleware class, that means user wants to proxy WebSocket connection, and configuration is OK, we shouldn't do anything here. 
But you can't open websocket directly, you should make a HTTP❗️ call with "Upgrade" header to webserver that handles websocket connections, then server responds with status "101 Switching Protocols" and then it [becomes a websocket](https://en.wikipedia.org/wiki/WebSocket#Protocol_handshake). 

"ws://" and "http://" in connection's URL are equal in this case (also "wss://" and "https://" accordingly), but passing user's "http" as protocol to [System.Net.WebSockets.ClientWebSocket.ConnectAsync()](https://learn.microsoft.com/en-us/dotnet/api/system.net.websockets.clientwebsocket.connectasync?view=net-7.0) will generate an exception "Only Uris starting with 'ws://' or 'wss://' are supported". So, we need to replace protocol in Url string before opening low-level WebSocket to handle proxy request. Replacing protocol [is common for constructing URL](https://stackoverflow.com/questions/10406930/how-to-construct-a-websocket-uri-relative-to-the-page-uri) from a given one, there is no magic here.